tools/src/main/java/org/apache/kafka/tools/ConnectPluginPath.java

```diff
@@ -0,0 +1,399 @@
+            } catch (UncheckedIOException e) {
+                throw new TerseException("Unable to parse plugin path " + rawPluginPath + ": " + e.getMessage());
+            }
+        }
+        for (String rawLocation : rawLocations) {
+            pluginLocations.add(Paths.get(rawLocation));
+        }
+        return pluginLocations;
+    }
+
+    enum Command {
+        LIST
+    }
+
+    private static class Config {
+        private final Command command;
+        private final Set<Path> locations;
+        private final PrintStream out;
+        private final PrintStream err;
+
+        private Config(Command command, Set<Path> locations, PrintStream out, PrintStream err) {
+            this.command = command;
+            this.locations = locations;
+            this.out = out;
+            this.err = err;
+        }
+
+        @Override
+        public String toString() {
+            return "Config{" +
+                "command=" + command +
+                ", locations=" + locations +
+                '}';
+        }
+    }
+
+    public static void runCommand(Config config) throws TerseException {
+        try {
+            for (Path pluginLocation : config.locations) {
+                ClassLoaderFactory factory = new ClassLoaderFactory();
+                try (DelegatingClassLoader delegatingClassLoader = factory.newDelegatingClassLoader(ConnectPluginPath.class.getClassLoader())) {
+                    Set<PluginSource> pluginSources = PluginUtils.pluginSources(Collections.singletonList(pluginLocation), delegatingClassLoader, factory);
+                    Predicate<PluginSource> isolated = PluginSource::isolated;
+                    PluginSource isolatedSource = pluginSources.stream()
+                            .filter(isolated)
+                            .findFirst()
+                            .orElseThrow(() -> new IllegalStateException("DelegatingClassLoader should have a PluginSource corresponding to" + pluginLocation));
+                    PluginSource classpathSource = pluginSources.stream()
+                            .filter(isolated.negate())
+                            .findFirst()
+                            .orElseThrow(() -> new IllegalStateException("DelegatingClassLoader should have a PluginSource corresponding to the classpath"));
+                    PluginScanResult scanResult = new ReflectionScanner().discoverPlugins(Collections.singleton(isolatedSource));
+                    PluginScanResult serviceLoadResult = new ServiceLoaderScanner().discoverPlugins(Collections.singleton(isolatedSource));
+                    PluginScanResult merged = new PluginScanResult(Arrays.asList(scanResult, serviceLoadResult));
+                    Set<ManifestEntry> isolatedManifests = findManifests(isolatedSource);
+                    Set<ManifestEntry> classpathManifests = findManifests(classpathSource);
+                    classpathManifests.forEach(isolatedManifests::remove);
```

Comment:
If a particular classname is included in two different manifest files (e.g. classpath and plugin.path) then their URIs will be different, as the URI includes the full path to the jar that contains the manifest file. That means that this remove() call will never have one remove the other. 

This remove() call is targeted at classpath manifests that are visible via the isolated classloader. When you call the PluginClassLoader#getResources(), that method will return both the isolated manifest and the classpath manifest as separate URLs. This will cause both isolatedManifests and classPath manifests to contain ManifestEntry objects with equivalent manifestURIs. Its the same problem that #13356 previously solved for the PluginDesc.